idx/memory/find_cache.go

```diff
@@ -0,0 +1,195 @@
+package memory
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/grafana/metrictank/stats"
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/raintank/schema"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	// metric idx.memory.find-cache.hit is a counter of findCache hits
+	findCacheHit = stats.NewCounterRate32("idx.memory.find-cache.hit")
+	// metric idx.memory.find-cache.miss is a counter of findCache misses
+	findCacheMiss = stats.NewCounterRate32("idx.memory.find-cache.miss")
+
+	findCacheSize            = 1000
+	findCacheInvalidateQueue = 100
+	findCacheBackoff         = time.Minute
+)
+
+type FindCache struct {
+	cache map[uint32]*lru.Cache
+	size  int
+	sync.RWMutex
+	newSeries map[uint32]chan struct{}
+	backoff   map[uint32]time.Time
+}
+
+func NewFindCache(size int) *FindCache {
+	fc := &FindCache{
+		cache:     make(map[uint32]*lru.Cache),
+		size:      size,
+		newSeries: make(map[uint32]chan struct{}),
+		backoff:   make(map[uint32]time.Time),
+	}
+	return fc
+}
+
+func (c *FindCache) Get(orgId uint32, pattern string) ([]*Node, bool) {
+	c.RLock()
+	cache, ok := c.cache[orgId]
+	c.RUnlock()
+	if !ok {
+		findCacheMiss.Inc()
+		return nil, ok
+	}
+	nodes, ok := cache.Get(pattern)
+	if !ok {
+		findCacheMiss.Inc()
+		return nil, ok
+	}
+	findCacheHit.Inc()
+	return nodes.([]*Node), ok
+}
+
+func (c *FindCache) Add(orgId uint32, pattern string, nodes []*Node) {
+	c.RLock()
+	cache, ok := c.cache[orgId]
+	t := c.backoff[orgId]
+	c.RUnlock()
+	var err error
+	if !ok {
+		// dont init the cache if we are in backoff mode.
+		if time.Until(t) > 0 {
+			return
+		}
+		cache, err = lru.New(c.size)
+		if err != nil {
+			log.Errorf("memory-idx: findCache failed to create lru. err=%s", err)
+			return
+		}
+		c.Lock()
+		c.cache[orgId] = cache
+		c.newSeries[orgId] = make(chan struct{}, findCacheInvalidateQueue)
```

Comment:
these newSeries chans smell a lot like util.Limiter, maybe we should use that to be more explicit ?

Reply:
util.Limiter is a blocking limiter.  We need a non-blocking one.